src/util/sharing_node.h

```diff
@@ -12,339 +12,455 @@ Author: Daniel Poetzl
 #ifndef CPROVER_UTIL_SHARING_NODE_H
 #define CPROVER_UTIL_SHARING_NODE_H
 
-#include <list>
+//#define DEBUG
+
+#ifdef DEBUG
+#include <iostream>
+#endif
+
+#include <forward_list>
+#include <type_traits>
+
+#ifndef SN_SMALL_MAP
+#define SN_SMALL_MAP 1
+#endif
+
+#ifndef SN_SHARE_KEYS
+#define SN_SHARE_KEYS 0
+#endif
+
+#if SN_SMALL_MAP == 1
+#include "small_map.h"
+#else
 #include <map>
-#include <memory>
+#endif
 
 #include "invariant.h"
+#include "make_unique.h"
+#include "small_shared_ptr.h"
+#include "small_shared_two_way_ptr.h"
+
+//#define SN_INTERNAL_CHECKS
+
+#ifdef SN_INTERNAL_CHECKS
+#define SN_ASSERT(b) INVARIANT(b, "Sharing node internal invariant")
+#define SN_ASSERT_USE(v, b) SN_ASSERT(b)
```

Comment:
In a previous comment it has been asserted that these utilities are not being used. Do you think they might be useful in the future, and if not, can they be removed?

Reply:
There were two similar macros, `SM_ASSERT_USE` and `SN_ASSERT_USE` (`SM` for sharing map and `SN` for sharing node). The former was unused and has now been removed, but the latter is still used.